### PR TITLE
log and return error on incorrect component type

### PR DESCRIPTION
**Pull request by mukundansundar**
# Description

log and return error on incorrect component type

additionally added some more unit tests for `doProcessComponent` and `initState`

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #2146 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/runtime/runtime.go [hunk 1]
```diff
@@ -1261,13 +1261,14 @@ func (a *DaprRuntime) appendOrReplaceComponents(component components_v1alpha1.Co
...
+func (a *DaprRuntime) extractComponentCategory(component components_v1alpha1.Component) (ComponentCategory, error) {
 	for _, category := range componentCategoriesNeedProcess {
 		if strings.HasPrefix(component.Spec.Type, fmt.Sprintf("%s.", category)) {
-			return category
+			return category, nil
 		}
 	}
-	return ""
+	// the category entered is incorrect, return error
+	return "", errors.Errorf("incorrect category %s", component.Spec.Type)
```
Comment (youngbupark) [inline pkg/runtime/runtime.go:1271]: `processComponentAndDependents` is the only method calling `extractComponentCategory`. I think we do not need to return error. Instead, `processComponentAndDependents` can check the return value is empty and then return the proper error.

what do you think?

File: pkg/runtime/runtime.go [hunk 2]
```diff
@@ -1294,7 +1295,11 @@ func (a *DaprRuntime) processComponentAndDependents(comp components_v1alpha1.Com
 		return nil
 	}
 
-	compCategory := a.extractComponentCategory(comp)
+	compCategory, err := a.extractComponentCategory(comp)
+	if err != nil {
```
Comment (youngbupark) [inline pkg/runtime/runtime.go:1299]: We can check if compCategory is empty or not. Then return the proper error.

**Review by youngbupark — CHANGES_REQUESTED** [review]

File: pkg/runtime/runtime.go [hunk 3]
```diff
@@ -1295,6 +1295,12 @@ func (a *DaprRuntime) processComponentAndDependents(comp components_v1alpha1.Com
 	}
 
 	compCategory := a.extractComponentCategory(comp)
+	if compCategory == "" {
+		// the category entered is incorrect, return error
+		err := errors.Errorf("incorrect type %s", comp.Spec.Type)
```
Comment (youngbupark) [inline pkg/runtime/runtime.go:1300]: we do not need to log error in this method (L1301 and L1305). Instead, we log error in caller code and reduce the duplicated code.
the below code is the only one caller part calling `processComponentAndDependents`. We can print out returned error. 

https://github.com/dapr/dapr/blob/cd35c75320423c7396fcd8bc6abfccdcc52427f4/pkg/runtime/runtime.go#L1274-L1279

**Review by youngbupark — CHANGES_REQUESTED** [review]

File: pkg/runtime/runtime.go (hunk 3 above)
Comment (mukundansundar) [reply pkg/runtime/runtime.go]: Good catch. Fixed.

**Comment by mukundansundar** [comment]
@youngbupark  please review this

**Review by youngbupark — APPROVED** [review]

**Pull request merged**